### ISSUE TITLE
XSLT in eXist: suggestions and problems

### ISSUE THREAD

**Issue by christian-schwaderer** [eXist-db/exist#791]
Dear eXist developers, dear all,

it certainly a big plus that you can make use of XSLT in eXist by having Saxon at hand. That makes in my case things a lot easier.

However, having used the transform:transform function now for some months, I encountered various issues, which I want to inform you about. Some I did already mention on the mailing list, but I considered it better to have them all in one place. (I don't know whether it is a good idea to dump that all in one issue. Feel free to re-organize the thread according to your taste.)

1.  Sometimes it would be handy to have the version of the XSTL processor at hand. A simply string function like “transform:version” returning something like “Saxon HE 9.5...” would helpful. This one should be easy to realise, I hope.

2. As XSLT transformations tend to throw errors, I wrapped the transform:transform call into a try-catch-block. However, it doesn't work as expected for all type of errors. Errors of the type: "exerr:ERROR XSL transform reported error:...." are 
treated as expected, while errors of the type "exerr:ERROR Exception while transforming node...” are not caught.

3. If there is an XSLT error, it is crucial to know where it is. In Oxygen, Saxon puts out the exact number of the line causing the problem. In eXist, there is none of that. Not even in the log files. Debugging becomes thus very painstaking for larger XSLT files. 

4. The output of XSL:message should not only be displayed in the console window, but should be passed through to the calling query. Maybe adding a function like transform:get-messages could be the solution?

5. This one is seemingly a bug: An XSLT transformation started in eXist returns for the XSLT command <xsl:value-of> not just text nodes but also the values of attribute nodes.

Let's say, we have an XML document named “simple_xml_doc.xml” of which the content is just
```
<root>
   <element attribute="attribute_value">Text content</element>
</root>
```

Then we use the following XSLT stylesheet:

```
<xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform" xmlns:xs="http://www.w3.org/2001/XMLSchema" exclude-result-prefixes="xs" version="2.0">
  <xsl:template match="/">
        <xsl:variable name="simple_xml_doc" select="doc('simple_xml_doc.xml')"/>
        <xsl:value-of select="$simple_xml_doc"/> 
    </xsl:template>   
</xsl:stylesheet>
```

We run that twice: Once in Oxygen, using Saxon-HE 9.6.0.5 and once in eXist 3.0 RC1 (You see, having a “get-version” function would be helpful, because I have no idea which Saxon version is used). The only other difference is that for the Oxyen run, “simple_xml_doc.xml” is a file on the file system, whereas in the case of eXist, it is stored in the database.

The output is different. In Oxyen, we get what we expect:

```
<?xml version="1.0" encoding="UTF-8"?>
   Text content
```

eXist gives us:

```

   attribute_valueText content

```

And that is NOT what we want to have. According to the XSLT specs, attribute values are not to be put out.

Just for completeness, the XQuery to start the transformation is
```
transform:transform(doc('input.xml'), doc('xslt.xsl'), (), (),())
```

The input file in both cases being just 
```
<root> foo! </root>
```

6. Finally, there is a performance issue. We set up a similar scenario as before. However, this time, we use a rather large XML file (about 44mb) and we extract an element 1 million times.

As XSLT, we use now:

```
<xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform" xmlns:xs="http://www.w3.org/2001/XMLSchema" exclude-result-prefixes="xs" version="2.0">
    <xsl:template match="/">
        <xsl:variable name="large_xml_doc" select="doc('large_xml_doc.xml')"/>
        <xsl:for-each select="1 to 1000000">
            <xsl:value-of select="$large_xml_doc//element[@another_attribute='special_element']"/> 
        </xsl:for-each>
    </xsl:template>   
</xsl:stylesheet>
```

In the file “large_xml_doc.xml”, there is somewhere an element 
```
<element attribute="value" another_attribute="special_element">It's me!</element>
```
Retrieving it 1 million times from a large file takes at least a measurable amount of time.

However, the combination Saxon/Oxygen/file system is way faster than Saxon/eXist/eXist storage.

Started within eXist it takes more or less exactly 16 seconds (according to “Query Profiling” in eXist). On the same machine, started with Oxygen it's about 1 second (according to Oxygen status bar)

**Comment by dizzzz** [comment]
Hmmm there has been no feedback on this issue for a long time. Sorry.
You make valuable suggestions! I am not sure if the more complex issues can be solved, because the xslt processor is an 3rd party library.  In the situation that somethings is behaving different, the only thing we probably can change is the XML serializer (which is somehow triggered by the XSLT:doc() function)

**Comment by adamretter** [comment]
So a few comments:

1) At the moment eXist abstracts the XSLT processor behind the JAXP API, this allows us to use any JAXP compatible library (e.g. Xalan or Saxon). This is really a legacy thing as eXist started with Xalan and then we added support for Saxon. The problem with using JAXP is it does not allow for a nice integration with Saxon, for example we can't get the detailed error information that you are asking for.

IMHO we should dump JAXP and the idea of a pluggable Transformer, and just settle on Saxon. In reality I think everyone uses Saxon and it is really the only modern and reliable Open Source choice in the Java world. Doing so we could use the S9 API which would allow us much better error reporting and integration.

2) eXist and Saxon have different DOM models, so each time we go from one to the other (i.e. via the XQuery transform:transform(...) function) we pay a serialization/deserialization cost. Given the time and funding, having talked with Michael Kay from Saxonica about the issue, it seems we could create a Navigable DOM integration for Saxon so that it operates directly on eXist's persistent DOM; this should remove the serialization/deserialization overhead. However, this is not a small piece of work and personally it is not a priority for me right now. If we could find the funding I would be in favour of asking Saxonica if they could undertake that work.

**Comment by dizzzz** [comment]
I agree, for the amount of work and the nature, finding some funding would be fair,

**Comment by duncdrum** [comment]
While this is ancient history, please add a 'needs documentation' label to this issue, to make it easier to track from the main documentation repo. 🙈 🙉 🙊
